Initial structure + some definitions. (#6)
dapr/sig-sdk-spec: spec/spec.md

# Dapr SDK Spec - Version 1.9-alpha

## Abstract

Dapr SDK Spec is a language-neutral specification serving as a reference to Dapr SDKs.

## Table of Contents

- [Overview](#overview)
- [Notations and Terminology](#notations-and-terminology)
- [Versioning](#versioning)
- [Client Object](#client-object)
- [Object Serialization](#object-serialization)
- [Exception Handling](#exception-handling)
- [Core Building Blocks](#core-building-blocks)
- [Actors](#Actors)
- [Preview APIs](#preview-apis)

## Overview

> [Dapr](https://github.com/dapr/dapr) offers a sidecar with an HTTP and gRPC API listening on localhost. In order to offer an additional layer of convenience, Dapr also offers SDKs in a variety of languages. Additionally, actors in Dapr rely on the SDK for applications to implement their actor classes and easy inocation of actor methods, timers and reminders.

Reviewer: ```suggestion
[Dapr](https://github.com/dapr/dapr) offers a sidecar with an HTTP and gRPC API listening on localhost. In order to offer an additional layer of convenience, Dapr also offers SDKs in a variety of languages. Additionally, actors in Dapr rely on the SDK for applications to implement their actor classes and innovation of actor methods, timers and reminders.
```

Author: I meant invocation.

Each SDK has its own set of maintainers and decisions are made per repository. Over time, decisions per SDK maintainer can deviate from a consistent experience when using Dapr in each programming language. This specification's goal is to bring standards to the Dapr experience via SDKs while still allowing each SDK to be idiomatic to its programming language and target audience.

## Notations and Terminology

### Notational Conventions

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).

### Terminology

This specification defines the following terms:

#### SDK

An "SDK" or "Software Development Kit" is defined as one or more libraries that offers a set of public interfaces, methods, classes and attributes offering a programming model for Dapr native to a given programming language.

#### Domain Object

A "domain object" is defined as an instance of a class from the application's code and not part of the SDK.

## Versioning

### Spec Versioning

> This spec follows [semantic versioning 2.0](https://semver.org/spec/v2.0.0.html). Matching the Dapr runtime API version for major and minor. It means that for each minor version release of Dapr runtime, there will be a new release of the spec. The patch version of the spec and runtime increase independently.

Reviewer 2: The subtlety of this is it reads (to me) as you need to enforce keeping version numbers the same between SDK and runtime.  But I believe, and correct me, this is also trying to state there needs to be 1 to 1 parity at all times between SDK and runtime.  Could we please clarify?

Author: I reworded this to make it clear that the parity is between the spec and the runtime. The SDKs can still be versioned independently - so the table in docs that matches the SDK versions with runtime versions remains and the spec compliance is automatically done via the runtime version.

The spec will be worked on in the main branch and include the `-alpha` suffix until it reached a stable version. Once the first stable version of the spec is reached, it will be under the release-x.y branch and tagged with vX.Y.0, such as the runtime release process.

This allows SDK developers to claim compliance towards a fixed spec and runtime version instead of a moving target.

### SDK Versioning

An SDK MUST start as one repository in the [dapr-sandbox organization](https://github.com/dapr-sandbox) at version 0.1, following the process to onboard a new Dapr sandbox project. Before an SDK can graduate to the [dapr organization](https://github.com/dapr), it MUST comply to a stable version of this spec, in addition to the graduation criteria for Dapr sandbox projects.

## Client Object

> The SDK MUST offer a way to instantiate a client object without requiring any parameter, using the `DAPR_HTTP_PORT` or `DAPR_GRPC_PORT` environment variables injected by the Dapr sidecar injector or Dapr CLI - defaulting to `3500` or `50001`, correspondingly, if environment variable is not set.

Reviewer 2: One more I feel passionate about is `APP_PORT` which leads to many mismatch bugs.

Author: The SDKs don't get this environment variable injected. This is in the sidecar.

> The client object MUST connect to the Dapr sidecar based on the `DAPR_SIDECAR_IP` environment variable, defaulting to `127.0.0.1` if not set.

Reviewer: Does this ENV Variable exist `DAPR_SIDECAR_IP`? If so this is not used in any samples and not documented. I doubt anyone is aware of this.

Author: It exists in the Java SDK. I can remove it from the spec until we agree to this.

The client object MUST abstract the protocol used to communicate to the Dapr sidecar and not offer gRPC or HTTP specific input or output.

> The client object MIGHT offer both gRPC or HTTP implementations.

Reviewer 2: I notice some of our SDKs leak a number of gRPC extensions, and then the API becomes hard to use if you don't know how to use it (meaning which APIs to use and which to ignore).

I believe this guidance will really help!

Reviewer: Really? Can you provide an example?

Author: I need some clarity here. I did not understand the issue.

> The client object SHOULD NOT implement any Actor functionality since that is handled as a separate construct.

Reviewer: I presume Actors are going to be covered later in this spec?

Author: Yes. There is a separate section. And Actors will be the hardest one, IMO.

The client object MUST implement all required methods for all [core building blocks](#core-building-blocks).

The client object MIGHT provide overloaded methods with reduced number of parameters for simpler method invocations.

> The client object SHOULD NOT implement any of the methods for the [preview APIs](#alpha-apis)

Reviewer: ```suggestion
The client object SHOULD NOT implement any of the methods for [alpha APIs](#alpha-apis)
```

> The client object SHOULD NOT implement any of the methods for the [preview APIs](#alpha-apis)

Reviewer 3: I see what we're suggesting for the alpha APIs below, but I don't think this is a good solution. See my comment there but I think we should do this differently if we're able to.

Author: I was expecting this to trigger a discussion since each SDK is doing something different here.

> ## Object Serialization

Reviewer 3: What about accepting raw bytes? This is a pretty common ask in the .NET SDK.

Author: OK. I will add that. We have that too. It is a special case of serialization since byte[] is usually an object.

Author: Added that statement to make it clear.

The SDK MUST serialize domain objects to JSON by default.

The SDK MIGHT accept a serializer implementation via a provided interface.

The SDK MIGHT accept a separate state serializer implementation via a provided interface to differentiate serialization of transient domain objects vs persisted domain objects.

## Exception Handling

Behavior and structure for exceptions is not defined yet.

## Core Building Blocks

* [State Store](core/state-store.md)
* [PubSub](core/pubsub.md)
* [Input Binding](core/input-binding.md)
* [Output Binding](core/output-binding.md)
* [Secrets Store](core/secrets-store.md)

## Actors

> An SDK MIGHT implement the Actors building block.

Reviewer 2: If we create new inter-dependencies I'd love for some ceremonious declaration of this.

Author: Do you mean compatibilities between Actor invocations cross languages?

Reviewer 4: Why is this MIGHT and not MUST?  What does not `MIGHT` mean? It's not part of the initial notational convention.

Author: The notation is defined above. I put MIGHT because Actors is kind of a separate thing from the "core APIs".

Author: I am changing it to MUST until we decide to make it optional. It seems to be the less controversial way to go now.

* [Actor client](actors/client.md)
* [Actor runtime](actors/runtime.md)

> ## Preview APIs

Reviewer: ```suggestion
## Alpha APIs
```

> The Preview APIs MUST be offered via a preview client object implementing a separate interface from the core's client object. The preview client object MUST implement the preview APIs below:

Reviewer: ```suggestion
The alpha APIs MUST be offered via an alpha client object implementing a separate interface from the core's client object. The alpha client object MUST implement the alpha APIs below:
```

> The Preview APIs MUST be offered via a preview client object implementing a separate interface from the core's client object. The preview client object MUST implement the preview APIs below:

Reviewer 3: I'm not a huge fan of this decision. The way we handle this in the .NET SDK is by using the `[Obsolete]` attribute. This attribute forces the caller to acknowledge that the API is not stable (compilation failure if you don't). 

Using a preview client also means that you'll end up with a breaking change when an API graduates to stable as you'll need to update your code to use the standard client. This could be avoided by having the preview client map over to the standard client but then you just have additional code paths.

Author: This is an interesting discussion and it shows the need for a spec. I don't think we should optimize for no breaking changes for preview APIs graduating since they will break at the API level regardless. I am not sure if [obsolete] is giving the right message either since the API is not obsolete, it is "experimental".

I think the solution should make it clear that the app is using an Alpha API. Eager to see what others think here. We did this PreviewClient in the Java SDK and it looked pretty clean from the core client. It also makes it clear that it does not have the same guarantees as the core client.

Reviewer 3: Yeah, I agree that `[Obsolete]` isn't the best attribute, but it's unfortunately the only one of that style in .NET. It does let you provide a message though, so when your code fails to compile it'll say something like "The ... API is current an Alpha API and is subject to change..."

The way we did it in the .NET SDK is that you won't need to have a breaking change when we change the API. The `DaprClient` is built as a layer over the gRPC provided client. So, when we update the proto, the methods will update. We can then call the new methods in the SDK and everything goes back to normal. When the user updates to the new revision, all they'll notice is that the `Obsolete` is gone but the compilation path remains the same. 

For languages that don't have support for this, it does become a bit harder. I wonder if we can do something else like an extension onto the client? You can import a separate package or project that contains those methods but they're labeled the same as the alpha APIs are an interface/inheritance/extension. When the graduate, they're moved to the base class and out of the alpha package. Thoughts?

Author: Interesting. If they are in a different package, then it would need to be a different interface, right? I am thinking of how to make the spec language agnostic. So, asking to have a different interface will accomplish that in a way that is more language neutral. In Java SDK, it is a separate interface. The class implements both interfaces but from the user's perspective, he is getting another type when we asks for an alpha client. I don't think we should optimize not to have a breaking change in SDK when an API graduates.

Reviewer 5: > The Preview APIs MUST be offered via a preview client object implementing a separate interface from the core's client object.

My feedback is that this will make it too hard for users to use preview features in practice. In a complex business application, objects like a Dapr client are threaded through several layers of glue code to get to where they are actually called. This puts a lot of burden on the user to update these references to refer to the new **preview** type, and then update them again in the future to refer to the stable one.  This kind of idea can work well for dynamic type systems, but it's also less meaningful when you've got a dynamic type system.

Of the problems SDKs/Libraries face, there are two typical problems with preview features:

- Users are not aware of the preview features and thus, they don't get a lot of feedback until they are stable.
- Users accidentally use preview features and have mismatched expectations.

In my experience the former problem is way more common, so I wouldn't want to put up even more barriers.

I'd suggest being less prescriptive and talking about the goal instead. Something like:

> SDKs SHOULD use an in-code mechanism in addition to documentation to communicate to users about the instability of preview features. For a function to interface with the new **widget** building block could be called `Preview_GetWidget()`

Author: Since this prescription seems to be controversial. I will remove it and allow us to have a separate discussion about it.

> * [Configuration](preview/configuration.md)

Reviewer: ```suggestion
* [Configuration](alpha/configuration.md)
```

> * [Distribured Lock](preview/distributed-lock.md)

Reviewer: ```suggestion
* [Distributed Lock](alpha/distributed-lock.md)
* [State Management Query](alpha/state-management.md)
```

- [Actors](#Actors)
- [Alpha APIs](#alpha-apis)

## Overview

[Dapr](https://github.com/dapr/dapr) offers a sidecar with an HTTP and gRPC API listening on localhost. In order to offer an additional layer of convenience, Dapr also offers SDKs in a variety of languages. Additionally, actors in Dapr rely on the SDK for applications to implement their actor classes and easy invocation of actor methods, timers and reminders.

Each SDK has its own set of maintainers and decisions are made per repository. Over time, decisions per SDK maintainer can deviate from a consistent experience when using Dapr in each programming language. This specification's goal is to bring standards to the Dapr experience via SDKs while still allowing each SDK to be idiomatic to its programming language and target audience.

## Notations and Terminology

### Notational Conventions

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).

### Terminology

This specification defines the following terms:

#### SDK

> An "SDK" or "Software Development Kit" is defined as one or more libraries that offers a set of public interfaces, methods, classes and attributes offering a programming model for Dapr native to a given programming language.

Reviewer 4: ```suggestion
An "SDK" or "Software Development Kit" is defined as one or more libraries that offers a set of public interfaces, methods, classes and attributes offering a programming model for Dapr, native to a given programming language.
```

The client object MUST connect to the Dapr sidecar on `127.0.0.1` and not `localhost` because some users have reported name resolution issues where `localhost` would not resolve to `127.0.0.1` or `::1`.

The client object MUST abstract the protocol used to communicate to the Dapr sidecar and not offer gRPC or HTTP specific input or output.

The client object MIGHT offer both gRPC or HTTP implementations.

The client object SHOULD NOT implement any Actor functionality since that is handled as a separate construct.

The client object MUST implement all required methods for all [core building blocks](#core-building-blocks).

The client object MIGHT provide overloaded methods with reduced number of parameters for simpler method invocations.

The client object SHOULD NOT implement any of the methods for the [alpha APIs](#alpha-apis)

## Object Serialization

The SDK MUST serialize domain objects to JSON by default, for both input and output.

> The SDK MUST accept `byte[]` as domain object for input and output.

Reviewer 5: IMO specifying the behavior of this requires a lot more thought. I'm not saying this is a bad scenario, but the expectations of what this does need to be clarifed. For example a lot of .NET JSON libraries will serialize `byte[]` as a base64-encoded JSON string. If the SDK author does the naive thing, then that would likely be the result.

Reviewer 5: My other feedback is that this is also too prescriptive. We should talk about the goal rather than the implementation. For example in .NET it's not really idiomatic to use `byte[]`. For large requests/responses this is very costly.

Someone designing an API for 'raw' I/O in .NET would probably start with `Stream` and then add `ReadOnlySpan<byte>` if there was demand for it.

Reviewer 4: I agree. I think should be more in terms of saying that the `SDK must accept domain objects serialized as bytes` or some such phrase rather than specifying any specific input/output format like a `byte[]`.

Author: I can reword this statement to not force the type.

There is a good point brought by Ryan of how much prescriptive the SDK spec should be. If we agree that the differences among SDKs right now are not a concern, that might remove the need for a spec at all (IMO). At the end of this spec, there will be changes required in each SDK to be compliant with the SPEC. I know that each maintainer made decisions in isolation and (sometimes) purposely different (like in "pluggable serializer"). I am targeting a spec that is minimum and removing controversial decisions for a later point.

As the spec evolves, there will be points where SDKs would need to make changes to be "compliant" to the spec.

> The SDK MIGHT accept a serializer implementation via a provided interface.

Reviewer 4: Should we clarify here that Actor serialization is different from other serialization and that it will use a completely different serializer?

Author: Everything in Actors is separate.
